src/coreclr/vm/eetwain.cpp

```diff
@@ -1693,60 +1660,84 @@ PTR_VOID EECodeManager::GetExactGenericsToken(PREGDISPLAY     pContext,
+    pCodeInfo->GetCodeManager()->EnsureCallerContextIsValid(pContext, NULL);
+
+    return EECodeManager::GetExactGenericsToken(GetSP(pContext->pCallerContext),
+                                                pCodeInfo,
+                                                GetIP(pContext->pCallerContext));
 }
 
 //static
 PTR_VOID EECodeManager::GetExactGenericsToken(SIZE_T          baseStackSlot,
-                                              EECodeInfo *    pCodeInfo)
+                                              EECodeInfo *    pCodeInfo,
+                                              UINT_PTR        returnAddress /* = 0 */)
 {
     LIMITED_METHOD_DAC_CONTRACT;
 
     GCInfoToken gcInfoToken = pCodeInfo->GetGCInfoToken();
 
     GcInfoDecoder gcInfoDecoder(
             gcInfoToken,
-            GcInfoDecoderFlags (DECODE_PSP_SYM | DECODE_GENERICS_INST_CONTEXT)
+            GcInfoDecoderFlags (DECODE_GENERICS_INST_CONTEXT)
             );
 
     INT32 spOffsetGenericsContext = gcInfoDecoder.GetGenericsInstContextStackSlot();
     if (spOffsetGenericsContext != NO_GENERICS_INST_CONTEXT)
     {
-
-        TADDR taSlot;
-        if (pCodeInfo->IsFunclet())
+        // For funclets we need to recover the establisher frame from
+        // the main function.
+        //
+        // Finally funclets can be called directly from the managed code
+        // in non-exceptional path. In that case the we don't need to do
+        // any adjustment.
+        //
+        // For calls made through CallEHFunclet and CallEHFilterFunclet
+        // we recover the establisher frame address from the frame of the
+        // helper call.
+        //
+        // Presumably profiler callbacks are not generated for funclets
+        // so we can use returnAddress == 0 and skip this.
+        if (returnAddress != 0 &&
+            pCodeInfo->IsFunclet() &&
+            !ExecutionManager::IsManagedCode(returnAddress))
         {
-            INT32 spOffsetPSPSym = gcInfoDecoder.GetPSPSymStackSlot();
-            _ASSERTE(spOffsetPSPSym != NO_PSP_SYM);
-
-#ifdef TARGET_AMD64
-            // On AMD64 the spOffsetPSPSym is relative to the "Initial SP": the stack
-            // pointer at the end of the prolog before and dynamic allocations, so it
-            // can be the same for funclets and the main function.
-            // However, we have a caller SP, so we need to convert
-            baseStackSlot -= pCodeInfo->GetFixedStackSize();
-
-#endif // TARGET_AMD64
-
-            // For funclets we have to do an extra dereference to get the PSPSym first.
-            TADDR newBaseStackSlot = *PTR_TADDR(baseStackSlot + spOffsetPSPSym);
+            SIZE_T offsetOfEstablisherFrameInFuncletSP;
+#if defined(TARGET_AMD64)
+            offsetOfEstablisherFrameInFuncletSP = 0;
+#elif defined(TARGET_X86)
+            // CallEHFunclet frame size + return address + 16 (5th parameter)
+#ifdef UNIX_X86_ABI
+            offsetOfEstablisherFrameInFuncletSP = 48;
```

Comment:
It would be good to add some static asserts to verify that these offsets are valid. I am worried that some future change could break these.

Reply:
I'm not "married" to the solution yet.

I tried to export the offset as variable from the ASM helper but it turns out that won't compile in DAC because the ASM helpers are excluded. It seems that the code could get called from some daccess.cpp code path so I abandoned that route. 

I'm not sure yet if I could easily turn this into static assert but I will definitely do it if I find a way.